coderd/dynamicparameters/render.go

```diff
@@ -131,46 +131,48 @@ func (r *loader) Renderer(ctx context.Context, db database.Store, cache *files.C
 		return r.staticRender(ctx, db)
 	}
 
-	return r.dynamicRenderer(ctx, db, cache)
+	return r.dynamicRenderer(ctx, db, files.NewCacheCloser(cache))
 }
 
 // Renderer caches all the necessary files when rendering a template version's
 // parameters. It must be closed after use to release the cached files.
-func (r *loader) dynamicRenderer(ctx context.Context, db database.Store, cache *files.Cache) (*dynamicRenderer, error) {
+func (r *loader) dynamicRenderer(ctx context.Context, db database.Store, cache *files.CacheCloser) (*dynamicRenderer, error) {
+	closeFiles := true // If the function returns with no error, this will toggle to false.
+	defer func() {
+		if closeFiles {
+			cache.Close()
+		}
+	}()
```

Comment:
oh I do not like this at all. this may be one of the very few times where I'd actually prefer something like named return values and then checking `err != nil`.

Reply:
I just saw this in another PR and actually really liked it.

https://github.com/coder/coder/pull/18448/commits/6d07d78574e281e51625dbb56c3f303222ac983e


By adding the boolean flip at the end of the function, it guarantees if the body exits early, it will close all the files. This is more future proof than a named return.

Reply:
I really wish Golang had some opinion on this style of code